Sources/NIOCore/ByteBuffer-hexdump.swift

```diff
@@ -0,0 +1,131 @@
+        if readableBytes <= limit {
+            return self.hexDumpShort()
+        } else {
+            let startHex = self.getSlice(at: 0, length: limit/2)!.hexDumpShort()
+            let endHex  = self.getSlice(at: self.readableBytes - limit/2, length: limit/2)!.hexDumpShort()
+            return startHex + " ... " + endHex
+        }
+    }
+
+    /// Returns a `String` of hexadecimal digits of bytes in the Buffer,
+    /// with formatting compatible with output of `hexdump -C`.
+    func hexDumpLong() -> String {
+        var result = ""
+
+        var buffer = self
+        var lineOffset = 0
+
+        while buffer.readableBytes > 0 {
+            let slice = buffer.readSlice(length: min(16, buffer.readableBytes))!
+            let lineLength = slice.readableBytes
+
+            result += "\(String(byte: lineOffset, padding: 8))  "
+
+            // Start with a xxd-format hexdump that we already have, then pad it
+            // and insert a space in the middle.
+            var lineHex = slice.hexDumpShort()
+            lineHex.append(String(repeating: " ", count: 49 - lineHex.count))
+            lineHex.insert(" ", at: lineHex.index(lineHex.startIndex, offsetBy: 24))
+            result += lineHex
+
+            // ASCII column renders the line as ASCII characters, or "." if the character is not printable.
+            result += "|" + String(decoding: slice.getBytes(at: 0, length: slice.readableBytes)!, as: Unicode.UTF8.self).map {
+                $0 >= " " && $0 < "~" ? $0 : "."
+            } + "|\n"
+
+            lineOffset += lineLength
+        }
+
+        // Add the last line with the last byte offset
+        result += String(byte: lineOffset, padding: 8)
+        return result
+    }
+
+//    func hexDumpLong(limit: Int) -> String {
+//        var result = ""
+//
+//        // Start with long-format dump of limit/2 bytes
+//        // then insert an empty line with "..." in the middle
+//        // Add a dump of the last limit/2 bytes
+//        // The last line should be the final lineOffset
+//
+//        
+//
+//        return result
+//    }
+
+    /// Describes a ByteBuffer hexDump format. Can be either xxd output compatible, or hexdump compatible.
+    /// You can provide a `maxLength` argument to both formats that will limit the maximum length of the buffer to dump before clipping the dump for readability.
+    public enum HexDumpFormat {
+        /// A hexdump format compatible with the xxd utility.
+        /// - parameters:
+        ///     - maxLength: The maximum amount of bytes to dump before clipping for readability.
+        case xxdCompatible(maxLength: Int? = nil)
+
+        /// A hexdump format compatible with hexdump utility.
+        /// - parameters:
+        ///     - maxLength: The maximum amount of bytes to dump before clipping for readability.
+        case hexDumpCompatible(maxLength: Int? = nil)
+    }
```

Comment:
Unfortunately we can't use an `enum` because adding a `case` is an API breaking change.

We can, however, have a struct which wraps an `enum`:


```swift
public struct HexDumpFormat {
    enum Value {
         case xxdCompatible(maxLength: Int?)
         case hexDumpCompatible(maxLength: Int?)
    }

    var value: Value
    init(_ value: Value) { self.value = value }

    public static let xxdCompatible = Self(.xxdCompatible(maxLength: nil))

    public static func xxdCompatible(maxLength: Int) -> Self { 
        Self(.xxdCompatible(maxLength: maxLength))
    }

    // etc...
}
```

Reply:
why can't we use enum again

Reply:
ahh thanks!

Reply:
I _think_ it's okay to switch `format.value` still in our code, right?

Reply:
yes, that's absolutely fine because the `Value` is `internal`